Add draft posts directory (#41)
sidneyw/sidneyw.com: projects/frontend/src/draft-posts/index.md

---
title: Go HTTP Reverse Proxy
author: Sidney
date: 1562204746
tags: ['go']
img: "./assets/header2.png"
---

> It’s no secret that the go has one of the most comprehensive standard libraries out there but did you know it had a reverse proxy implementation?

Reviewer: ```suggestion
It’s no secret that go has one of the most comprehensive standard libraries out there, but did you know it had a reverse proxy implementation?
```

> Go is one of my favorite programming languages for a many reasons. Simplicity, practicality, and performance were key focus areas for the designers of the language. Where this truly shines is with networking tasks.

Reviewer: ```suggestion
Go is one of my favorite programming languages for many reasons. Simplicity, practicality, and performance were key focus areas for the designers of the language. Where this truly shines is with networking tasks.
```

This post will show how simple it is to set up a reverse proxy in go and extend the proxy to capture metrics from downstream services.

> 

Reviewer: One thing about this entire section is that I'm still not sure what you're telling me about or why. When I leave the intro, I should know:
1. Who you are
2. Why you're writing this
3. What you're writing about.

As it is, I'm not sure who you are or why you're writing it. Also, I think your descriptions of Go in the first and second paragraphs should probably be collapsed into one collection of thoughts. Keeping your previous theme, something along the lines of "Go is great. This is why. Also, reverse proxy isn't that cool?"

## Reverse Proxying

> If you’re new to the concept of reverse proxying an exact definition would be:

Reviewer: I don't have a suggestion, but this sentence feels awkward. Re-write it if you can, but otherwise it's probably fine.

> ”In [computer networks](https://en.wikipedia.org/wiki/Computer_network) , a reverse proxy is a type of [proxy server](https://en.wikipedia.org/wiki/Proxy_server) that retrieves resources on behalf of a client from one or more servers. These resources are then returned to the client, appearing as if they originated from the proxy server itself. “
>
> From [Wikipedia](https://en.wikipedia.org/wiki/Reverse_proxy)

Essentially, a reverse proxy forwards traffic from a client to a set of servers behind the proxy.

![Reverse Proxy Network](./assets/proxy.png)

> Why would you want to do this? There are many applications for reverse proxies. Load balancing, TLS termination, metric reporting, and A/B testing are just a few.

Reviewer: I feel like you should answer this question a lot earlier. Like in the "why am I writing this" portion. Specifically, there are two `whys` that I think are collapsed into one. 
1. Why would one want a reverse proxy?
2. Why would one want to implement their own instead of using nginx or something as a reverse proxy?

If you’d like to learn more about proxying I highly recommend checking out  [Introduction to modern network load balancing and proxying](https://blog.envoyproxy.io/introduction-to-modern-network-load-balancing-and-proxying-a57f6ff80236)  by Matt Klein. Matt is the creator of [Envoy Proxy](https://www.envoyproxy.io/), a robust proxy server that powers service mesh tools like [Istio](https://istio.io/). His post does a great job of outlining the approaches used by modern load balancers and proxies.

## Simple Go Reverse Proxy
How do we go about rolling our own simple proxy in go?

```go
proxy := httputil.NewSingleHostReverseProxy(url)
```

Yep, with the go standard library it really is that simple.

> Let’s dig in here. The function returns a reverse proxy with the following method.

Reviewer: Which function? :)

```go
func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
```

> All we need to do is configure the proxy and wire it up to a standard go http server to have a working reverse proxy as shown below.

Reviewer: I would totally link to the godocs for ReverseProxy somewhere in here.

```go
package main

import (
	"flag"
	"log"
	"net/http"
	"net/http/httputil"
	"net/url"
	"strconv"
)

func main() {
	port := flag.Int("port", 8080, "port to listen on")
	targetURL := flag.String("target-url", "", "downstream service url to proxy to")
	flag.Parse()

	u, err := url.Parse(*targetURL)
	if err != nil {
		log.Fatalf("Could not parse downstream url: %s", *targetURL)
	}

	proxy := httputil.NewSingleHostReverseProxy(u)

	director := proxy.Director
	proxy.Director = func(req *http.Request) {
		director(req)
		req.Header.Set("X-Forwarded-Host", req.Header.Get("Host"))
		req.Host = req.URL.Host
	}

	http.HandleFunc("/", proxy.ServeHTTP)
	log.Printf("Listening on port %d", *port)
	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(*port), nil))
}
```

> That’s it! This server can now handle proxying HTTP 1 and 2 as well as web socket connections. You’ll notice that I’ve configured the _proxy.Director_ field. The _ReverseProxy.Director_ is a function that modifies the incoming request before it is forwarded. The signature is as follows:

Reviewer: I'm not sure what exactly will fix it, but it took me 3 tries to mentally parse `HTTP 1 and 2` as `HTTP versions 1 and 2` and not `Two arbitrary http servers which I have named 1 and 2`

Reviewer: Also, maybe this is just a go thing, but I like signature and then code and not code and then signature. But that's personal preference and worth very little.

```go
Director func(*http.Request)
```

> A common use case for the director function is altering request headers. There is a default director set by  _httputil.NewSingleHostReverseProxy_ . The default implementation takes care of setting the request scheme, host, and path. I found this useful and didn’t want to duplicate the code. I had to reset the _req.Host_  field to contact HTTPS endpoints . I’ve also included an example of setting a request header.

Reviewer: Given the go "Useful by default" mentality, I might say something like this instead:
```suggestion
A common use case for the director function is modifying request headers. The default director implementation for _httputil.NewSingleHostReverseProxy_ takes care of setting the request Scheme, Gost, and Path. I found this useful and didn’t want to duplicate the code, so I wrapped that implementation. *Note*: I had to reset the _req.Host_  field to handle HTTPS endpoints. I’ve also included an example of setting a request header.
```

Reviewer: Two questions:
1. What are the implications of overwriting the `req.Host` property?
2. Does `req.Header.Set` overwrite or append? This is important as I found out at work a couple months ago and I would mention it.

## Capturing Metrics 

> Not too long ago I was asked to build a demo for one of our clients at work. The task was to capture the response metrics of a web service and add those metrics to [Prometheus](https://prometheus.io/) for display in [Grafana](https://grafana.com/). The only catch was that I did not have direct access to the web service and therefore couldn’t instrument the code. I completed the task by adapting the go reverse proxy to observe the network traffic coming from the service.

Reviewer: This is the explanation of why that should be started a lot earlier.

Also, this is the place where people typically plug their company 😃

Reviewer: I just suggested you split this up and bring some of the "why" earlier, but if you don't, I'd suggest an edit of this section. It doesn't feel like it has a direct enough relationship to the previous section. Something along the lines of: 
```markdown
One way reverse proxies are useful is they allow you to insert instrumentation around an http service without having to modify the service itself.
``` 
followed by a slight tweak of your paragraph. Also, it's an implementation of the decorator pattern, but your HTTP service is the function!

> Let’s extend our simple proxy to read and report metrics about the downstream service responses. To do this we’ll return to the  _httputil.ReverseProxy_ struct once more. It exposes a  struct field _ReverseProxy.ModifyResponse_ which gives us access to the http response before it goes back to the client.

Reviewer: ```suggestion
Let’s extend our simple proxy to read and report metrics about the downstream service responses. To do this we’ll return to the  _httputil.ReverseProxy_ struct once more. It exposes a struct field _ReverseProxy.ModifyResponse_ which gives us access to the http response before it goes back to the client.
```

Reviewer: One a separate note. HTTP or http. Please pick one :).

```go
ModifyResponse func(*net/http.Response) error
```

> HTTP bodies in go are implemented as _io.Reader_’s and therefore they can only be read once. If you would like to parse them before forwarding the request/response you will need to copy the body into a byte buffer and the reset the request/response body. An obvious drawback is that we buffer the entire response in memory without limit. This could lead to memory issues in production if you received a large response but for our use case this wasn’t an issue. Here’s a quick implementation to duplicate the response body.

Reviewer: ```suggestion
HTTP bodies in go are implemented as _io.Reader_s and therefore can only be read once. If you would like to parse them before forwarding the HTTP message you will need to copy the body into a byte buffer and the reset the body. An obvious drawback is that we buffer the entire response in memory without limit. This could lead to memory issues in production if you received a large response but for our use case this wasn’t an issue. Here’s a quick implementation to duplicate the response body.
```

Reviewer: I would also specify at this point that there's no reason to actually read the body unless you actually need something out of it.

```go
func duplicateReader(in io.ReadCloser) (b, b1 io.ReadCloser, err error) {
	bBytes, err := ioutil.ReadAll(in)
	if err != nil {
		return nil, nil, err
	}

	in.Close()

	buf := bytes.NewBuffer(bBytes)
	b = ioutil.NopCloser(buf)
	b1 = ioutil.NopCloser(bytes.NewBuffer(bBytes))

	return
}
```

> Depending on your use case you may not actually want to duplicate the reader. You may opt for returning the parsed body of the response i.e. parsing the json into a struct or _map[string]interface{}_  and just resetting _req.Body_ to an equivalent reader.

Reviewer: I would potentially go through the trouble of moving this to the end and giving that implementation. Or leaving it commented in the code snippet

With the request body problem solved, capturing metrics is simple.

```go
	proxy := httputil.NewSingleHostReverseProxy(u)
	// Configure proxy.Director ...

	// ModifyResponse runs before forwarding the downstream response back to the client
	proxy.ModifyResponse = func(res *http.Response) error {
			var err error
			var body io.ReadCloser

			body, res.Body, err = duplicateReader(res.Body)
			if err != nil {
				panic(err)
			}

			fmt.Printf("body = %+v\n", body)

			captureMetrics(body)
			return nil
		}
}
```

> And that’s it! The capture metrics function was pretty specific to my use case so I’ll leave it up to you to implement. Form here it is pretty straightforward to further extend for applications like load balancing, redirect policies, etc.

Reviewer: ```suggestion
And that’s it! The capture metrics function was specific to my use case so I’ll leave it up to you to implement. From here it is pretty straightforward to further extend for applications like load balancing, redirect policies, etc.
```

Reviewer: Also, I do not believe it is simple to extend for load balancing. You implemented a proxy with a single backend. You should link to a multi-backend standard library version if available.

> Hopefully you’ve left this post as excited as I am about the go standard library for networking tasks. If you found this at all useful or have any questions please leave a comment below.

Reviewer: Can I get the full program now? :)

Reviewer: I'm not sure what to do about it, but your closer feels weak. Again, a super common thing for tech blogs, but it's worth mentioning.
